@using EuroConnector.ClientApp.Data.Interfaces;
@using EuroConnector.ClientApp.Data.Models;

@inject ISetupService SetupService
@inject ISnackbar Snackbar
@inject NavigationManager Navigation
@inject ILogger Logger

@attribute [AllowAnonymous]

<div>
    <MudText Typo="Typo.h5">Login</MudText>

    <MudForm @ref="form" @bind-IsValid="success">
        <EconTextField T="string" @bind-Value="@properties.UserName" Required="true" Label="Username" />
        <EconTextField T="string" @bind-Value="@properties.SecretKey" Required="true" InputType="InputType.Password" Label="Secret Key"/>
        <EconTextField T="string" @bind-Value="@properties.ApiUrl" Required="true" Label="API URL"/>
    </MudForm>

    <div class="button-group">
        <EconButton OnClick="OnSaveClicked" Disabled="!success">Save</EconButton>
        <EconButton OnClick="OnClearClicked">Logout</EconButton>
    </div>
</div>

@code {
    [CascadingParameter]
    public Shared.Error Error { get; set; }

    bool success;
    MudForm form;
    LoginSettings properties = new();

    protected override async Task OnInitializedAsync()
    {
        properties = await SetupService.GetLoginSettings();
    }

    private async Task OnSaveClicked()
    {
        Logger.Information("Login save clicked.");

        try
        {
            await SetupService.Login(properties);
            Snackbar.Add("Login settings applied successfully.", Severity.Success);
            Logger.Information("Login settings applied successfully.");
            Navigation.NavigateTo("/");
        }
        catch (Exception ex)
        {
            Error.ProcessError(ex);
            Logger.Error(ex, "Login exception.");
        }
    }

    private async Task OnClearClicked()
    {
        Logger.Information("Logout clicked.");

        try
        {
            await SetupService.Logout();
            Snackbar.Add("Logged out successfully.", Severity.Success);
            Logger.Information("Logged out successfully.");
        }
        catch (Exception ex)
        {
            Error.ProcessError(ex);
            Logger.Error(ex, "Logout exception.");
        }
    }
}
